sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql
+
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.test.SharedSQLContext
+
+case class T1(a: Int, b: Option[Int], c: Option[Int])
+
+/**
+ * This test suite takes https://sqlite.org/nulls.html as a reference.
+ */
+class NullHandlingSuite extends QueryTest with SharedSQLContext {
```

Comment:
My original proposal is to do it in `SQLQueryTestSuite.scala` like the other `.sql` files. Just curious why you want to write it using Dataset APIs?

Reply:
I wasn't aware of your original proposal, sorry.
I simply prefer it because I think it is less error prone.

I can change it if you want.

Reply:
When you create a .sql file, you just need to run the command:
> SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"

Why you think it is error-prone?

Reply:
I think that in general writing SQL code is more error prone than writing Scala code because there are no compile checks and I prefer to write Scala code than SQL for this reason. Moreover, I think it is easier to do some things like running for loops over all the possible 0's implementations. With SQL I should have copied and pasted the same query multiple times to achieve the same. But maybe this is not even necessary, we can just check one.

As I said, I wasn't aware that you had a proposal for the location of the checks, thus I just did what I considered the best option. If you want, I can move them to a sql file.

Reply:
Maybe using SQL? It will be easier for the others who knew SQL only to understand our NULL handling logics. Thanks again!